Haste2

cpojer opened the pull request:
This pull request is used to track the implementation of node-haste2.

##### Summary
* Implemented node-haste2, a re-implementation of Facebook's haste dependency system, based on Facebook's react-native packager. We now create the entire haste-map upfront and pass it to the worker threads instead of creating node-haste instances in each worker.
* Changed the core architecture of Jest to abstract the HasteModuleLoader into a ModuleLoader and HasteResolver piece. Added a Test class so we aren't creating instances of a TestRunner in every worker. The new code should bring a lot of clarity but there is also a lot more to be done.
* Performance improvements
 * fixed startup time of Jest, both on cold and warm start. node_modules are now lazily analyzed.
 * fixed require-resolution time for relative requires. Loading modules without haste is now much faster and uses more caching.
* New implementation of `--onlyChanged` that should work much better and find the correct set of related tests to run.
* Numerous bugfixes and cleanups all around Jest.

##### Performance Improvements
* The Relay test suite (GitHub) now takes 20s vs. 60-100s previously (on 13” MBP)
* The react-native test suite now takes 8 seconds instead of ~70s (on 13” MBP)
* Starting up Jest on react-native (fb internal) now takes 1 second vs. ~10-15 seconds from before.
* FB's frontend codebase test run was already heavily optimized but there is ~10 % win down to ~97s from about 107s.

This diff/PR forms the basis for all the work on Jest that I'm planning to do in the near term and will set up Jest for further success. It should be much easier to iterate on Jest from now on given that we have replaced the biggest legacy system and I can finally focus on solving some of the community problems.

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
The new module execution code is a little bit more ugly but it is super-hot code and yields a 10 sec perf win on www (~8.5 % of the entire runtime).

cpojer commented:
@facebook-github-bot import

cpojer commented:
omg it works.

browniefed commented:
#dope

cpojer commented:
Outstanding list of TODOs:
* Add cache and fileWatcher to actually make this fast
* Only construct one moduleLoader in TestRunner; re-use HasteResolver
* Make 'HasteResolver' class configurable
* Test this at FB and fix issues that come up

Future release:
* Support NODE_PATH + traversal up the tree for node_modules so we can drop `resolve`

None of these should be blockers for merging this PR except for testing this at FB (no point in building this if it doesn't work on all of our tests ). I will try to run it on Monday and fix minor issues as they come up (assuming it won't blow up completely :) ). The first for TODO items are blockers for tagging a new version of jest itself, of course.

jsdf commented:
Correct me if I'm wrong, but it looks like the (soon to be pluggable) module resolver still can only provide filepaths to resolved modules, and the file contents are actually loaded from disk by the `lib/transform` module.

From the perspective of being able to plug-in different module transformation+bundling pipelines (eg. webpack, browserify), with the goal of reducing the amount of duplicated config/machinery required to use Jest, it would be great to have the option of providing module sources as the output of the resolver, to be used, instead of only providing filepaths.

In both browserify and webpack, resolving and transforming of modules is interleaved, and transforms/loaders are async, so the easiest place to plug them in would probably be in place of `HasteResolver`, but also providing the transformed module source text along with the resolved module filepath.

jsdf commented:
All the other stuff looks good though :)

cpojer commented:
@jsdf thanks for the review! What would the ideal API look like to you? I take it webpack/browserify pre-transform code and then resolve them?

I can imagine to merge the preprocessor and the resolver: resolving the dependencies for a path will pre-transform the file and apply the transform recursively for every dependency before resolving its dependencies. Now that there is only a single entry point for transform (there used to be three!) this should be easy to change. All that it needs is a map from real path to the cache path, which should be achievable using the `getCacheKey` method in the preprocessor.

cpojer commented:
High level description of all the changes:

* replaced node-haste with node-haste2
* major cleanups in HasteModuleLoader, TestRunner and `jest.js`
* fixed a bunch of bugs.
* Split up HasteModuleLoader into two classes, the new class is HasteResolver which will soon receive more functionality.

Here is a breakdown of all the changes from oldest to latest:

https://github.com/cpojer/jest/compare/89481fbf1a7d7ce9b37b85bb6ffbd0a76cda69ad...cpojer:5f82cc75872ae02e97fcfa5f41039f8f6d6e3686
* The first few commits are cleanups. They remove some cruft in the test runner and inline the module execution. By not using objects and `Object.keys` when executing a module we get a slight speedup. Note that this code will move out of HasteModuleLoader in the future and I just consolidated this convoluted piece of code

https://github.com/cpojer/jest/commit/e1faf1bd8ea38cb14c8f2cc955e47685782ef788
* This diff is inconsequential. `getJestRuntimeForTest` was removed in a later commit in the stack.

https://github.com/cpojer/jest/commit/620ea9e003751e96bbee139f139e61a71043665d
* This diff cleans up a bunch of minor things. Most notably I noticed that the `disposed` variable wasn't useful at all. I'm now using the global to check if a context should be disposed.

https://github.com/cpojer/jest/commit/f518a9235f4273af64559dde9cb7e047c0fc0f14
* Slightly unrelated change but we were reporting the wrong number of total test suites (we didn't include the failed ones, this was annoying during the refactors)

https://github.com/cpojer/jest/compare/9a4e8d0a11f8055f5683c9fc7d8ee6a712fbb263...cpojer:000d0d9d6c53b64e5c25b9e2018c3035c8f50a4b
* This is the collection of diffs with all the changes that replace node-haste with node-haste2. At the end I changed all objects in the loader from `{}` to `Object.create(null)`.

https://github.com/cpojer/jest/commit/422ab8863d02953682f6fcfe1f762297a4c9d25d
* This diff reorganizes all the methods in HasteModuleLoader, the file is now readable. I sorted it similar to how classes are organized in node-haste, with the public interface on top. The one change I made was to not only create coverage data for direct dependents. This doesn't seem useful and isn't worth mutating the config object over. I can bring this back in the future if we need it but I think ignoring `node_modules` is enough. I froze the config object in the next diff: https://github.com/cpojer/jest/commit/1b06d9e9d9d78032568b3b0fbcb50b0e7d2a9a82 which will allow further optimizations in the future.

https://github.com/cpojer/jest/commit/92bcc173dc9936305ea409255621d6e8a951c57a
* Just a code cleanup and moved the code to ES2015

https://github.com/cpojer/jest/commit/5fd0c8888d4413b418c5bea6793772f04cee9d44
* I noticed `jest --onlyChanged` never properly worked in the jest repo itself. This fixes the command for repos where `<rootDir>` doesn't equal the git root. The big algorithmic change in this diff is that instead of looking at all the changed files and resolving their dependents recursively up to a test file, we are now looking at all the recursive dependencies of all the test files and intersect them with the changed files. Let's bring back the old algorithm (and support for it in node-haste) if this proves to be too slow internally at FB.

jsdf commented:
@cpojer your description of the combined preprocessor and resolver sounds right. For the use case I have in mind, the ideal API would be (most of which is already there):
- The module resolver class should be instantiated once and reused between test files.
- There should be a `getDependencies` method which is called with a test file before it is run by the test runner. This method should return a promise of a dependency map for that test file.
- the resulting dependency map can include preprocessed source contents. 
  In this case, either the Jest preprocessor implementation wouldn't be used, or alternatively if it is used by the custom module resolver implementation then there should be a version of the `transform` function which would be able to take `(src, filename)` as arguments rather than loading the source file from disk itself.

cpojer commented:
Ok we are back in business. I just rebased this branch on top of all the work that went into jest since I started on this diff. Everything up until "Use `Object.create(null)` in HasteResolver." was reviewed by @amasad about two weeks ago. I'll pull in the minor cleanups from the beginning of this PR into jest 0.8.1 and once I'm done with the jasmine2 conversion at FB I'll pick this one back up.

cpojer commented:
Another rebase with all of the haste2 unrelated changes removed and pushed to jest.

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
@facebook-github-bot import

cpojer commented:
I updated the PR summary with a better description of what is being accomplished here.

vitalets commented:
> I updated the PR summary with a better description of what is being accomplished here.

thanks a lot! When you plan to get it merged?

cpojer commented:
tomorrow!

vitalets commented:
wow! :+1:

irvinebroque commented:
😍👍🏾👏🏽🔥

irudoy commented:
🤘

cpojer commented:
Seeing similar perf gains in open source, from #jest on our discord channel from @ncuillery:
![Imgur](http://i.imgur.com/6Yj0qHa.png)

If you'd like to try this with react-native, you might need to specify the `haste` option in jest now:

```js
  "haste": {
      "defaultPlatform": "ios",
      "providesModuleNodeModules": [
        "fbjs",
        "react",
        "react-native",
        "parse",
        "react-transform-hmr"
      ],
      "platforms": ["ios", "android"]
    },
```

I'm not entirely sure whether putting `react-native` into `providesModuleNodeModules` is going to be sufficient. I'll do more testing before the release to make sure that apps dependent on react-native have a smooth upgrade experience.

cpojer commented:
Diff was landed against Facebook's main codebase and it seems to be working fine. I'll hold off merging this diff here for another day to make sure everything stays green internally.

cpojer commented:
Another person tried and got significant speedups:
![Imgur](http://i.imgur.com/KbrCB0J.png)

craigdallimore commented:
We've tried this on Windows 8 with a different outcome:

Using jest via `npm test`:
```
Using Jest CLI v0.9.0, jasmine1
No tests found for "./src/js".
```

And with an absolute path:
```
Using Jest CLI v0.9.0, jasmine1
No tests found for "C:/path/to/project/src/js".
```

This is also present with a globally installed `jest-cli`, at this version.

The same config is working on OSX. I'll do some more investigation on Windows when I get home and add details if I find anything.

ncuillery commented:
As the execution time of huge tests suites is far better, single test runs seem pretty slow now. 

I did some measurements here https://gist.github.com/ncuillery/0e0dbce66a9281de6fde

cpojer commented:
Thanks @ncuillery that is super useful. I haven't been able to repro this on any platform so far.

Would you mind sharing the size of your repository? How many files are visible from all the `testPathDirs` (or `rootDir` if `testPathDirs` wasn't specified) and how many of those are js/json? One thing you could try is to limit the `testPathDirs` to an array of folders as close to the JS as possible. At FB, for example, we have `/html/js` which is only a tiny subset of the root folder of all files in our repository. Crawling the root would be super slow, but crawling only the subset that contains JS is reasonably fast.

The new implementation uses watchman for file crawling if available and that should be much faster. Would you mind trying to set this up for your project? https://facebook.github.io/watchman/ I'm curious to hear about the difference.

One important thing to note is that I'm planning to improve the `--watch` command soon and to turn Jest into a server, so that single test runs are instant and there won't be a startup time penalty.

ncuillery commented:
You are right ! I had a problem with my config !

Sources are located in `<rootDir>/src` with ~50 sub-directories, all tests are located in `<rootDir>/tests/units` which contains the same structure (we didn't follow the default: having relative sub-directories  `__tests__`, an old choice we deeply regret today).  
`<rootDir>/src` contains 209 `js` files and 135 `jsx` files (React components). `<rootDir>/tests/units` contains 189 test files.

My config was containing `"testDirectoryName": "units"`. I added `"testPathDirs": ["<rootDir>/tests"]` and now the "warm-up" time is unnoticeable. 

Jest was looking for `units` directory in each sub-directories of `<rootDir>` including `<rootDir>/src` and its sub-directories (and also `<rootDir>/node_modules` ? not sure about that).

The whole test suite is now fast as hell: ~50s (and not ~60s like I said before)

cpojer commented:
Awesome!

Alright, it is time to merge. We should figure out the Windows issue before tagging 0.9.0.

cpojer commented:
I've published this PR as `0.9.0-fb1` for anyone who would like to try it out.

SomniumDigital commented:
Gone from 70 tests in 26 tests suites running in ~21secs to ~9secs (after warming up)
Great work @cpojer go and get yourself a beer 👍🍻😀

iamdustan commented:
:clap::clap:

irvinebroque commented:
Working great for us, also appears to have fixed https://github.com/facebook/jest/issues/632 :clap:

irudoy commented:
I'm running tests with `jest.runCLI()` via Gulp, and it still takes a lot of time. Also, when tests have been passed, gulp task doesn't exit and loading CPU at 80%. If I run it with `npm test`, all is ok. @cpojer

cpojer commented:
@irudoy can you try passing `watchman: false` to your jest run?

irudoy commented:
@cpojer it didn't help

cpojer commented:
If you are willing to add tests to Jest that point out this issue you are seeing with your integration or sahre a test suite with a repro, I can try to look into this.

irudoy commented:
@cpojer https://github.com/irudoy/jest/tree/master/examples/gulp

cpojer commented:
Thanks! Note that 0.9 is not out yet and you should be safe using 0.8. Would you mind opening a separate issue on this repo with the details of this issue and what you'd expect to work that doesn't any more? :)

irudoy commented:
@cpojer okay, here it is #722 :)

cpojer commented:
Published jest@next again as 0.9.0-fb2 which should resolve the watcher issues.

jrubins commented:
Performance improvements from `v0.8.2` to `v0.9.0` were SIGNIFICANT for us.

The following is output from a Jenkins build with `v0.8.2`:
<img width="522" alt="screen shot 2016-03-10 at 5 40 18 pm" src="https://cloud.githubusercontent.com/assets/971647/13688420/69e0faaa-e6e7-11e5-9d65-ed5c5d0a7342.png">

And the following is output from a Jenkins build after upgrading to `v0.9.0`:
<img width="523" alt="screen shot 2016-03-10 at 5 40 50 pm" src="https://cloud.githubusercontent.com/assets/971647/13688427/7ad856d2-e6e7-11e5-95a4-ff00058a2f5d.png">

_(Note: Coverage was turned on for both runs. The upgrade was added along with additional tests in a feature branch which is why the test suite numbers went up.)_

Really appreciate the dedication to improving the performance and look forward to future improvements!

cpojer commented:
Ohhh, there is an open task to improve caching when coverage is being used. If anyone wants to work on that, it should improve test runs with coverage by 20-50 %.
